<launch>

	<arg name="uavname" default="hil" />
    <arg name="imu_topic" default="/djiros/imu"/>
    <arg name="topic_odom_uav" default="/odom/uav"/>

    <remap from="/uwb_vicon_odom" to="$(arg topic_odom_uav)"/>

    <node pkg="djiros" type="djiros_node" name="djiros" output="log" launch-prefix="" required="true">
        <param name="log_level" type="string" value="info"/>
        <param name="serial_name" type="string" value="/dev/tty.dji.n3"/>
        <param name="baud_rate" type="int" value="921600"/>
        <param name="app_id"        type="int"      value="1024795"/>
        <param name="app_version"   type="int"      value="1"/>
        <param name="enc_key"       type="string"   value="a5f11a418f3dcdc39c4a1efb35a9dd4545e47b56cae9d98c988468277a9305dc"/>
        <param name="sensor_mode" type="bool" value="false"/>
        <param name="align_with_fmu" type="bool" value="false"/>
    </node>

    <node pkg="n3ctrl" type="n3ctrl_node" name="n3ctrl" output="log" required="true">
        <!--subscribers-->
		<remap from="~joy" to="/djiros/rc" />
        <remap from="~odom" to="$(arg topic_odom_uav)" />
		<remap from="~imu" to="/djiros/imu" />
		<remap from="~cmd" to="/position_cmd" />
        <!--publishers-->
		<remap from="~desire_pose" to="/desire_pose" />
		<remap from="~traj_start_trigger" to="/traj_start_trigger" />
		<remap from="~ctrl" to="/djiros/ctrl" />
        <!--params-->
        <rosparam command="load" file="$(find n3ctrl)/config/ctrl_param_$(arg uavname).yaml" />
        <param name="work_mode" value="real-time" />
        <param name="js_ctrl_mode" value="feedback" />
        <param name="pub_debug_msgs" type="bool" value="true"/>
    </node>

    <node pkg="joy" type="joy_node" name="joy_node" output="screen">
        <param name="~autorepeat_rate" value="0" />
    </node>

    <node pkg="rnw_ros" type="gamepad_trigger_node" name="gamepad" output="screen"/>

    <!-- odom for the drones, simulated and fake ones   -->

    <node pkg="rnw_ros" type="fake_drone_node" name="fake_drone1" output="screen" required="true">
        <remap from="~poly_traj" to="/drone1/traj"/>
        <remap from="~odom" to="/drone1/odom"/>
        <remap from="~position_cmd" to="/drone1/position_cmd"/>
        <remap from="~n3ctrl" to="/drone1/state"/>
    </node>

    <node pkg="n3ctrl" type="n1_sim_helper.py" name="n1_sim_helper" output="screen" required="true">
        <!--subscribers-->
        <remap from="~velo" to="/djiros/velo" />
        <remap from="~gps" to="/djiros/gps" />
        <remap from="~imu" to="/djiros/imu" />
        <!--publishers-->
        <remap from="~odom" to="/drone2/odom" />
    </node>

    <!--  generate control cmd for the drones  -->

    <node pkg="rnw_ros" type="test_swarm_planner_node" name="test_swarm_planner_node" output="screen"/>

    <node pkg="rnw_ros" type="traj2cmd_node" name="traj2cmd_drone1" output="screen" required="true">
        <remap from="~traj" to="/drone1/traj"/>
        <remap from="~odom" to="/drone1/odom"/>
        <remap from="~position_cmd" to="/drone1/position_cmd"/>
    </node>

    <node pkg="rnw_ros" type="traj2cmd_node" name="traj2cmd_drone2" output="screen" required="true">
        <remap from="~traj" to="/drone2/traj"/>
        <remap from="~odom" to="/drone2/odom"/>
        <remap from="~position_cmd" to="/drone2/position_cmd"/>
    </node>

    <!--  UWB in the loop  -->

    <node pkg="uwb_transceiver" type="uwb_transceiver_node" name="uwb_master_drone2" output="screen">
        <rosparam command="load" file="$(find uwb_transceiver)/config/config.yaml" />
        <param name="serial_port" value="/dev/tty.uwb.master2"/>
        <remap from="~odom" to="/drone2/odom"/>
        <remap from="~position_cmd" to="/drone2/position_cmd"/>
        <remap from="~n3ctrl" to="/drone2/state"/>
        <param name="role" type="int" value="1"/> <!--master-->
        <param name="debug/m_debug_packet_info" type="int" value="0"/>
    </node>

    <node pkg="uwb_transceiver" type="uwb_transceiver_node" name="uwb_slave_drone2" output="screen">
        <rosparam command="load" file="$(find uwb_transceiver)/config/config.yaml" />
        <param name="serial_port" value="/dev/tty.uwb.slave"/>
        <remap from="~odom" to="$(arg topic_odom_uav)"/>
        <remap from="~position_cmd" to="/position_cmd"/>
        <remap from="~n3ctrl" to="/n3ctrl/n3ctrl_state"/>
        <param name="role" type="int" value="2"/> <!--slave-->
        <param name="debug/m_debug_packet_info" type="int" value="0"/>
    </node>

    <!--  visualization  -->

    <node name="rviz_traj" pkg="rviz" type="rviz" output="log" args="-d $(find rnw_ros)/config/swarm.rviz" />

    <node pkg="rnw_ros" type="visualize_pos_cmd_node" name="sim_cmd" output="screen" required="true">
        <remap from="~position_cmd" to="/position_cmd"/>
    </node>

    <node pkg="rnw_ros" type="odom2tf_node" name="sim_tf" output="screen">
        <remap from="~odom" to="$(arg topic_odom_uav)"/>
        <param name="name" value="sim_tf" />
    </node>

    <node pkg="rnw_ros" type="odom2tf_node" name="drone1_tf" output="screen">
        <remap from="~odom" to="/drone1/odom"/>
        <param name="name" value="odom_drone1" />
    </node>

    <node pkg="rnw_ros" type="visualize_traj_node" name="traj_viz_drone1" output="screen" required="true">
        <param name="lift_dt" type="double" value="0.05"/>
        <param name="length_g" type="double" value="0.5"/>
        <param name="clear_after_n_sec" type="double" value="0"/>
        <remap from="~poly_traj" to="/drone1/traj"/>
    </node>

    <node pkg="rnw_ros" type="visualize_traj_node" name="traj_viz_drone2" output="screen" required="true">
        <param name="lift_dt" type="double" value="0.05"/>
        <param name="length_g" type="double" value="0.5"/>
        <param name="clear_after_n_sec" type="double" value="0"/>
        <remap from="~poly_traj" to="/drone2/traj"/>
    </node>

</launch>
